### PR TITLE
Add storage tests for Azure File in-tree storage plugin

### LINKED ISSUE #106677: Missing Azure File e2e tests
### What happened?

Azure File is one of the intree storage plugins available in k8s, however, we currently don't have a e2e tests for it.

Having such tests would be helpful to test CSI migration for Azure File.

### What did you expect to happen?

Azure File e2e tests should run on Azure.

### How can we reproduce it (as minimally and precisely as possible)?

Just run the storage e2e test suite on an Azure cluster.

### Anything else we need to know?

_No response_

### Kubernetes version

1.23

### Cloud provider

Azure

### OS version

_No response_

### Install tools

_No response_

### Container runtime (CRI) and and version (if applicable)

_No response_

### Related plugins (CNI, CSI, ...) and versions (if applicable)

_No response_

### PR DESCRIPTION
**Pull request by bertinatto**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This PR adds Azure File to the storage e2e test suite.

Another approach would be to use the external testsuite using a manifest file [like this one](https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/master/test/external-e2e/testdriver-smb.yaml). However, the external testsuite isn't ready for that (plus IMO tests that test intree plugins should be in k/k).

#### Which issue(s) this PR fixes:
Fixes #106677

#### Special notes for your reviewer:

This is work-in-progress. There are around 16 tests still failing.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by bertinatto** [comment]
/hold
There are around 16 failing tests that needs to be investigated before merging this.

**Comment by bertinatto** [comment]
Current status:

```
Summarizing 15 Failures:

[Fail] [sig-storage] In-tree Volumes [Driver: azure-file] [Testpattern: Dynamic PV (default fs)] subPath [It] should unmount if pod is gracefully deleted while kubelet is down [Disruptive][Slow][LinuxOnly]
/home/fbertina/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/utils/utils.go:151

[Fail] [sig-storage] In-tree Volumes [Driver: azure-file] [Testpattern: Pre-provisioned PV (filesystem volmode)] volumeMode [It] should not mount / map unused volumes in a pod [LinuxOnly]
/home/fbertina/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/framework/volume_resource.go:263

[Fail] [sig-storage] In-tree Volumes [Driver: azure-file] [Testpattern: Dynamic PV (filesystem volmode)] disruptive[Disruptive][LinuxOnly] [It] Should test that pv written before kubelet restart is readable after restart.
/home/fbertina/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/utils/pod.go:114

[Fail] [sig-storage] In-tree Volumes [Driver: azure-file] [Testpattern: Dynamic PV (default fs)] fsgroupchangepolicy [It] (Always)[LinuxOnly], pod created with an initial fsgroup, volume contents ownership changed in first pod, new pod with same fsgroup applied to the volume contents
/home/fbertina/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/utils/utils.go:728

[Fail] [sig-storage] In-tree Volumes [Driver: azure-file] [Testpattern: Inline-volume (default fs)] volumes [It] should store data
/home/fbertina/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/volumes.go:186

[Fail] [sig-storage] In-tree Volumes [Driver: azure-file] [Testpattern: Dynamic PV (default fs)] fsgroupchangepolicy [It] (Always)[LinuxOnly], pod created with an initial fsgroup, volume contents ownership changed in first pod, new pod with different fsgroup applied to the volume contents
/home/fbertina/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/utils/utils.go:728

[Fail] [sig-storage] In-tree Volumes [Driver: azure-file] [Testpattern: Pre-provisioned PV (default fs)] volumes [It] should store data
/home/fbertina/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/framework/volume_resource.go:263

[Fail] [sig-storage] In-tree Volumes [Driver: azure-file] [Testpattern: Inline-volume (default fs)] volumes [It] should allow exec of files on the volume
/home/fbertina/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/pods.go:99

[Fail] [sig-storage] In-tree Volumes [Driver: azure-file] [Testpattern: Dynamic PV (default fs)] fsgroupchangepolicy [It] (OnRootMismatch)[LinuxOnly], pod created with an initial fsgroup, volume contents ownership changed in first pod, new pod with same fsgroup skips ownership changes to the volume contents
/home/fbertina/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/utils/utils.go:728

[Fail] [sig-storage] In-tree Volumes [Driver: azure-file] [Testpattern: Pre-provisioned PV (default fs)] volumes [It] should allow exec of files on the volume
/home/fbertina/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/framework/volume_resource.go:263

[Fail] [sig-storage] In-tree Volumes [Driver: azure-file] [Testpattern: Dynamic PV (default fs)] volumeIO [It] should write files of various sizes, verify size, validate content [Slow]
/home/fbertina/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/volume_io.go:163

[Fail] [sig-storage] In-tree Volumes [Driver: azure-file] [Testpattern: Dynamic PV (default fs)] subPath [It] should be able to unmount after the subpath directory is deleted [LinuxOnly]
/home/fbertina/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/subpath.go:183

[Fail] [sig-storage] In-tree Volumes [Driver: azure-file] [Testpattern: Dynamic PV (default fs)] subPath [It] should unmount if pod is force deleted while kubelet is down [Disruptive][Slow][LinuxOnly]
/home/fbertina/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/utils/utils.go:151

[Fail] [sig-storage] In-tree Volumes [Driver: azure-file] [Testpattern: Dynamic PV (default fs)] fsgroupchangepolicy [It] (OnRootMismatch)[LinuxOnly], pod created with an initial fsgroup, volume contents ownership changed in first pod, new pod with different fsgroup applied to the volume contents
/home/fbertina/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/utils/utils.go:728

[Fail] [sig-storage] In-tree Volumes [Driver: azure-file] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode [It] should fail to create pod by failing to mount volume [Slow]
/home/fbertina/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/volumemode.go:209

Ran 50 of 7281 Specs in 4374.676 seconds
FAIL! -- 35 Passed | 15 Failed | 0 Pending | 7231 Skipped
```

**Comment by dims** [comment]
this PR has been on hold for about 2 weeks. Can you please see how to unblock this by reviewing comments etc? or please close this out (Trying to declutter our open list of PRs)

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml
Comment (enj) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml:995]: I would not expect RBAC changes due to test enhancements.

Comment (andyzhangx) [reply plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml:995]: secrets `["get", "list", "create"]` is required by azure file driver since it needs to create k8s secret to store account key and get secret when mount, and secrets `delete` is not required.

**Comment by andyzhangx** [comment]
about the above test errors:
 - `fsgroupchangepolicy` actually does not apply for azure file driver since gid is determined during smb mount (so we could ignore these 4 errors)
 - `subPath` issue was fixed on 1.21.9, 1.22.6, 1.23.2, related to https://github.com/kubernetes/kubernetes/pull/106906 ( 3 subPath errors)
 - `block volmode` is not supported for azure file driver (1 block volume mode error)

So we only got 7 test errors left, shall we merge this PR first?

**Comment by bertinatto** [comment]
/hold cancel
@andyzhangx thanks for reviewing!

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml
Comment (bertinatto) [reply plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml]: Thanks! Fixed

**Comment by andyzhangx** [comment]
> /retest Quite a few jobs are failing, but they seem unrelated to this change.

@bertinatto yep, I have checked, the apiserver does not work in those testing clusters, already asking azure capz team to investigate those failing cluster set up in tests.

**Comment by jsafrane** [comment]
It's very odd that `[It] should store data` tests won't pass - they're the simplest tests out there. Do you have complete output of the tests?

**Comment by andyzhangx** [comment]
@bertinatto could you follow this guide: https://github.com/kubernetes/kubernetes/pull/106678#issuecomment-1056511046, and what's the k8s version you are running on?

**Comment by bertinatto** [comment]
@andyzhangx I made a few more adjustments and now all tests pass!

```
$ _output/bin/e2e.test --provider=azure --cloud-config-file=azure_ocp.json --ginkgo.focus="In-tree Volumes \[Driver: azure-file\].+" --ginkgo.skip="\[Disruptive\]"

(...)

Ran 42 of 7371 Specs in 2022.031 seconds
SUCCESS! -- 42 Passed | 0 Failed | 0 Pending | 7329 Skipped
PASS
```

**Comment by andyzhangx** [comment]
```
running golangci-lint 
test/e2e/storage/drivers/in_tree.go:2035:2: field `resourceGroup` is unused (unused)
	resourceGroup   string
```

**Comment by bertinatto** [comment]
> ```
> running golangci-lint 
> test/e2e/storage/drivers/in_tree.go:2035:2: field `resourceGroup` is unused (unused)
> 	resourceGroup   string
> ```

Fixed

**Review by andyzhangx — APPROVED** [review]
/lgtm
/approve

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go
Comment (liggitt) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:272]: list permissions on secrets are not expected for the volume binder, only individual gets.

likewise, the volume binder should not be creating secrets

**Comment by bertinatto** [comment]
/hold cancel
dropped permissions for volume binder

**Review by neolit123 — COMMENTED** [review]
/approve
For the framework changes. I don't have a lot of context here so defering to sig storage and cloud provider. Ideally we should remove these provider specific tests from k/k...

Also there are a number of failing tests.
Remove this hold if this is good to go.
/hold

**Pull request merged**
